src/lesson/lesson04.js

```
@@ -0,0 +1,1173 @@
+      return new Lesson04Game.Error(
+        Lesson04Game.Error.COLLECTED_FROM_NO_SOURCE);
+    if (this.holding_item !== null)
+      return new Lesson04Game.Error(
+        Lesson04Game.Error.COLLECTED_WHILE_HOLDING_ITEM);
+    var item = source.pop();
+    if (item instanceof SourceEnd)
+      return new Lesson04Game.Error(
+        Lesson04Game.Error.COLLECTED_FROM_EMPTY_SOURCE);
+    if (/SHIP/.test(item))
+      return new Lesson04Game.Error(
+        Lesson04Game.Error.COLLECTED_FROM_NO_SOURCE);
+    this.holding_item = item;
+    return null;
+  },
+  // If arm is currently standing by some source, return what would be
+  // collected from it. Return null otherwise.
+  peek: function() {
+    var source = this.inverted_object_index[this.arm_pos];
+    if (!(source instanceof Source || source instanceof Machine))
+      return null;
+    return source.peek();
+  },
+  // Deposit item into the position the arm is currently standing by, returning
+  // null if successful (and transferring the item to the arm) of some error
+  // otherwise.
+  deposit: function() {
+    var deposit = this.inverted_object_index[this.arm_pos];
+    if (!(deposit instanceof Deposit || deposit instanceof Machine))
+      return new Lesson04Game.Error(
+        Lesson04Game.Error.DEPOSITED_ON_INVALID_LOCATION);
+    var result = undefined;
+    if (deposit instanceof Deposit)
+      result = deposit.push(this.holding_item);
+    else
+      result = deposit.push(this.arm_pos, this.holding_item);
+    if (result === null)
+      this.holding_item = null;
+    return result;
+  },
+  // Returns true if the game goal has been reached.
+  reached_goal: function() {
+    switch (this.goal) {
+      case Lesson04Game.Goals.EVERY_MACHINE_SOURCE_NOT_EMPTY:
+        for (var i = 0; i < this.machines.length; ++i)
+          if (this.machines[i].peek() instanceof SourceEnd)
+            return false;
+        return true;
+      case Lesson04Game.Goals.FILL_EVERY_DEPOSIT:
+      default:
```

Same thing as the other default: I think you can just throw an error string.